From: Carlos Garnacho <carlosg@gnome.org>
Date: Wed, 7 Aug 2024 13:59:20 +0200
Subject: wayland: Also cancel XDnD drops with no selected target
MIME-Version: 1.0
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: 8bit

This is usually an indication of a failed drop operation if the
drop site didn't request any target. Check for this specifically
on XDnD button release, so that we can cancel the DnD operation
right away.

Inspired on a fix from Jonas Ådahl.

Part-of: <https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/3821>

Origin: upstream, https://gitlab.gnome.org/GNOME/mutter/-/commit/ef3de1e58e96bb464f49e7ce6dde64c68bb2c71f
Bug: https://gitlab.gnome.org/GNOME/mutter/-/issues/3511
Bug-Ubuntu: https://launchpad.net/bugs/2097415
---
 src/wayland/meta-xwayland-dnd.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/src/wayland/meta-xwayland-dnd.c b/src/wayland/meta-xwayland-dnd.c
index 0a768eb..81aaeff 100644
--- a/src/wayland/meta-xwayland-dnd.c
+++ b/src/wayland/meta-xwayland-dnd.c
@@ -915,9 +915,10 @@ drag_xgrab_release (MetaWaylandEventHandler *handler,
                            CLUTTER_BUTTON3_MASK |
                            CLUTTER_BUTTON4_MASK |
                            CLUTTER_BUTTON5_MASK)) <= 1 &&
-      (!meta_wayland_drag_grab_get_focus (drag_grab) ||
-       meta_wayland_data_source_get_current_action (data_source) ==
-       WL_DATA_DEVICE_MANAGER_DND_ACTION_NONE))
+      ((!meta_wayland_drag_grab_get_focus (drag_grab) ||
+        meta_wayland_data_source_get_current_action (data_source) ==
+        WL_DATA_DEVICE_MANAGER_DND_ACTION_NONE) ||
+       !meta_wayland_data_source_has_target (data_source)))
     meta_xwayland_end_dnd_grab (&seat->data_device, FALSE);
 
   return CLUTTER_EVENT_PROPAGATE;
